common/crypto/openssl/cert.c

```diff
@@ -1019,3 +1033,333 @@ oe_result_t oe_cert_get_validity_dates(
+
+    X509_EXTENSION_free(ext);
+    ext = NULL;
+
+    /* Set the subject key idenfitifer extension */
+    ext = X509V3_EXT_conf_nid(NULL, &ctx, NID_subject_key_identifier, "hash");
+    if (!ext)
+        OE_RAISE_MSG(
+            OE_CRYPTO_ERROR, "create subject key identifier extension failed");
+
+    ret = X509_add_ext(x509cert, ext, -1);
+    if (!ret)
+        OE_RAISE_MSG(OE_CRYPTO_ERROR, "add basic constraint extension failed");
+
+    X509_EXTENSION_free(ext);
+    ext = NULL;
+
+    /* Set the authority key idenfitifer extension */
+    ext = X509V3_EXT_conf_nid(
+        NULL, &ctx, NID_authority_key_identifier, "keyid:always");
+    if (!ext)
+        OE_RAISE_MSG(
+            OE_CRYPTO_ERROR, "create subject key identifier extension failed");
+
+    ret = X509_add_ext(x509cert, ext, -1);
+    if (!ret)
+        OE_RAISE_MSG(OE_CRYPTO_ERROR, "add basic constraint extension failed");
+    X509_EXTENSION_free(ext);
+    ext = NULL;
+
+    /* Set the custom extension */
+    data = ASN1_OCTET_STRING_new();
+
+    ret = ASN1_OCTET_STRING_set(
+        data,
+        (const unsigned char*)config->ext_data_buf,
+        (int)config->ext_data_buf_size);
+    if (!ret)
+        OE_RAISE_MSG(OE_CRYPTO_ERROR, "set octet string failed");
+
+    obj = OBJ_txt2obj((char*)config->ext_oid, 1);
+    if (!obj)
+        OE_RAISE_MSG(OE_CRYPTO_ERROR, "create custom extension obj failed");
+
+    if (!X509_EXTENSION_create_by_OBJ(&ext, obj, 0, data))
+        OE_RAISE_MSG(OE_CRYPTO_ERROR, "create custom extension failed");
+
+    ret = X509_add_ext(x509cert, ext, -1);
+    if (!ret)
+        OE_RAISE_MSG(OE_CRYPTO_ERROR, "add custom extension failed");
+
+    /* Sign the certificate */
+    if (!X509_sign(x509cert, subject_issuer_key_pair, EVP_sha256()))
+        OE_RAISE_MSG(OE_CRYPTO_ERROR, "sign cert failed");
+
+    /*
+     * Write to DER
+     * The use of temporary variable is mandatory.
+     * If p is not NULL, the i2d_x509 function writes the DER encoded data to
+     * the buffer at *p and increments p to point after the data just written.
+     */
+    p = buff;
+    len = i2d_X509(x509cert, &p);
+    if (len <= 0)
+        OE_RAISE_MSG(OE_CRYPTO_ERROR, "i2d_X509 failed");
+
+    *bytes_written = (size_t)len;
+
+    /* Copy DER data to buffer */
+    OE_CHECK(oe_memcpy_s(
+        (void*)cert_buf, cert_buf_size, (const void*)buff, *bytes_written));
+    OE_TRACE_VERBOSE("bytes_written = 0x%x", (unsigned int)*bytes_written);
+
+    result = OE_OK;
+
+done:
+    if (x509cert)
+        X509_free(x509cert);
+    if (ext)
+        X509_EXTENSION_free(ext);
+    if (name)
+        X509_NAME_free(name);
+    if (bio)
+        BIO_free(bio);
+    if (obj)
+        ASN1_OBJECT_free(obj);
+    if (data)
+        ASN1_OCTET_STRING_free(data);
+    if (bc)
+        BASIC_CONSTRAINTS_free(bc);
+    if (subject_issuer_key_pair)
+        EVP_PKEY_free(subject_issuer_key_pair);
+    if (buff)
```

Comment:
Suggestion: free() is safe to call with NULL.  Delete unnecessary if tests, to simplify code.